express.js

```diff
@@ -537,14 +568,24 @@ module.exports = (app) => {
 		const instance = createApiWidgetInstanceData(data[0])[0];
 		instance.id = id;
 		instance.name = data[1];
+
 		fs.writeFileSync(path.join(qsets, id + '.instance.json'), JSON.stringify([instance]));
 
 		// send a warning back to the creator if any nonstandard question properties were detected
 		if (nonstandard_props.length > 0) {
 			const plurals = nonstandard_props.length > 1 ? ['properties', 'were'] : ['property', 'was'];
-			instance.warning = 'Warning: Nonstandard qset item ' +
+			console.log ('Warning: Nonstandard qset item ' +
 				plurals[0] + ' ' + nonstandard_props.join(', ') + ' ' +
-				plurals[1] + ' not saved. Use options instead.';
+				plurals[1]);
```

Comment:
IMO, perhaps it's valuable to include the caveat that the `options` object is the place to contain nonstandard qset items? Or is that taken to be a given based on documentation?

Reply:
There's nothing technically wrong with using 'nonstandard properties' on qset items, so in that regard it's allowed - I think we discussed requiring those to be stored in `options` instead, but the docs currently don't say it's strictly required to follow the suggested standard.

Ideally we would expect qset items to have a specific set of properties, one of which being 'options', and anything outside that specific set would be stored in 'options'. For widgets (like Enigma) that don't conform to the standard, this will require the widget to version its qsets and interpret all of the versions properly.

Reply:
Materia will strip non-standard properties from structures it determines to be 'questions', but there is some detection involved in that. Specifically:

```
foreach ($source as $key => &$q)
{
	//...
	if (self::is_question($q))
	{
		//...
		$real_q = Widget_Question::forge()->from_json($json);

//...

public static function is_question($object)
{
	$array = (array) $object;
	if ( ! array_key_exists('id', $array)) return false ;
	if ( ! array_key_exists('type', $array)) return false;
	if ( ! array_key_exists('questions', $array)) return false;
	if ( ! array_key_exists('answers', $array)) return false;
	if (empty($array['type']) || empty($array['questions']) || empty($array['answers']) ) return false;
	if ( ! is_array($array['answers'])) return false;
	if ( ! is_array($array['questions'])) return false;
	return true;
}
```

In the case of Enigma, again, questions are nested two layers deep: `qset->data->items->items`. Objects at that second `items` layer are determined to be 'questions' by Materia and so are converted to objects that have a specific set of properties, implicitly stripping everything else. Objects at the first layer, `qset->data->items` do not qualify as 'questions' so Materia does not strip any nonstandard properties from them.